Fix issue #4060 Add clang15/clang_trunk as HLSL compilers

thewilsonator opened the pull request:
N.B I don't really know what I'm here, just copying vaguely what #3961 did, and what is in `etc/config/c.amazon.properties`. What is the `.defaults.properties` and what does it do? do I need to update it too?

cc @llvm-beanz What options should I be setting here? I could add `-triple dxil-pc-shadermodel6.3-library` but that would preclude writing non-library dxil modules wouldn't it?

RubenRBS commented:
Thanks a lot for taking the time to open the PR.
> What is the .defaults.properties and what does it do? do I need to update it too?

We use a hierarchical config system where `.defaults` is always loaded for each language, so users that just clone & make have some usable defaults to use, and then we can overwrite it in production with the `.amazon` level. So yes it might be useful to add an entry in the defaults hier, for something like `/usr/bin/clang++-15` as c++.defaults does.

RubenRBS reviewed (changes requested):
Again, thank you so much for taking the time to make the necessary changes here :)

thewilsonator commented:
Thank you for the speedy review!

RubenRBS commented on etc/config/hlsl.defaults.properties:
```diff
@@ -10,6 +10,10 @@ group.rga.compilerType=rga
 compiler.rga_default.dxcPath=/usr/dxc-artifacts/bin/dxc
 
+group.clang.compilers=hlsl_clang_default
+group.rga.compilerType=clang
```
Typo, `group.rga`

RubenRBS replied:
Also should have the -x hlsl option set here too

thewilsonator replied:
Thanks, fixed.

llvm-beanz commented:
This change gets the HLSL language flags working through the clang driver directly:
https://reviews.llvm.org/D133958

thewilsonator commented:
>  I wouldn't add clang-15 for HLSL.

Removed.

> The -x hlsl flag shouldn't be needed as long as the source file has the .hlsl extension

Noted, removed.

> We also might need to force the -emit-llvm flag for now too

That works for me with what I want to use this for.

> The DirectX and SPIR-V backends aren't fully implemented and are probably not included in the compiler explorer toolchain.

Awesome, I will go under the assumption this lands soon and leave `--driver-mode=dxc` off.

thewilsonator commented:
Thanks!

mattgodbolt commented:
Hi @thewilsonator  - I tried this in our staging environment but I get:

```
error: HLSL code generation is unsupported for target 'x86_64-unknown-linux-gnu'
Compiler returned: 1
```

I'm going to push this to our production environment anyway as I don't know how to fix this and perhaps by you taking a look we can find out what I'm doing wrong

llvm-beanz commented:
HLSL requires targeting dxil-pc-shadermodel... something, but since that last something matters the user needs to provide it. For example:

dxil-pc-shadermodel6.0-compute
Or
dxil-pc-shadermodel5.0-pixel

The former specifies compute shaders with the full availability of shader model 6.0 APIs. The later targets 5.0 and pixel shaders.

mattgodbolt commented:
Ok - that gets further: https://godbolt.org/z/4EfK1sM8M

looks like `float4` isn't a thing (our default example uses it)...I don't know enough about the ecosystem (clearly) to know if that's "ok" or not

llvm-beanz commented:
The missing `float4` is caused by a change I haven't yet merged (https://reviews.llvm.org/D133958). I will merge it today.

The default shader is going to have other issues for a while because upstream clang's support has been focused on compute shaders, and the default shader is pixel.

thewilsonator commented:
https://reviews.llvm.org/D133958 was merged, `float4` should work the next time llvm-trunk is rebuilt.
